modules/shared/src/main/scala/org/tessellation/currency/utils/FeeTransactionUtils.scala

```diff
@@ -0,0 +1,58 @@
+package org.tessellation.currency.utils
+
+import cats.Applicative
+import cats.data.NonEmptyList
+import cats.effect.Async
+import cats.syntax.all._
+
+import org.tessellation.currency.dataApplication.{DataTransaction, DataUpdate, FeeTransaction}
+import org.tessellation.security.signature.Signed
+
+import derevo.circe.magnolia.{decoder, encoder}
+import derevo.derive
+import io.circe.{Decoder, Encoder}
+import org.http4s._
+import org.http4s.circe.jsonOf
+
+object FeeTransactionUtils {
+  def decoderFeeTransactionRequest[F[_]: Async, D <: DataUpdate](
+    implicit dataUpdateEncoder: Encoder[Signed[D]],
+    dataUpdateDecoder: Decoder[Signed[D]],
+    feeTransactionEncoder: Encoder[Signed[FeeTransaction]],
+    feeTransactionDecoder: Decoder[Signed[FeeTransaction]]
+  ): EntityDecoder[F, NonEmptyList[Signed[DataTransaction]]] = {
+    @derive(decoder, encoder)
+    case class DataTransactionRequest(
+      data: Signed[D],
+      fee: Option[Signed[FeeTransaction]]
+    )
+
+    implicit def dataTransactionRequestDecoder: EntityDecoder[F, DataTransactionRequest] =
+      jsonOf[F, DataTransactionRequest]
+
+    EntityDecoder.decodeBy(MediaType.application.json) { msg =>
+      DecodeResult {
+        msg
+          .as[DataTransactionRequest]
+          .flatMap { req =>
+            val signedTransactions: List[Signed[DataTransaction]] = List(req.data) ++ req.fee.toList
+            NonEmptyList.fromList(signedTransactions) match {
+              case Some(nonEmptyTransactions) =>
+                Applicative[F].pure(Right(nonEmptyTransactions): Either[DecodeFailure, NonEmptyList[Signed[DataTransaction]]])
+              case None =>
+                Applicative[F].pure(
+                  Left(InvalidMessageBodyFailure("Empty list of transactions")): Either[DecodeFailure, NonEmptyList[
```

Comment:
instead of `Invalid..(""): Either..`, use `.widen` or use a correct type on `Left`. Anyway, you can use `.asLeft` as well, so it would be:

```scala
Invalid..("errormsg").widen[..].asLeft[..].pure[F]
```

Reply:
👍🏼